### PR TITLE
[SOFT-415] Can Datagram 

### PR DESCRIPTION
**Pull request by mitchellostler**
Brief description of how Can Datagram works (see wiki [here](https://uwmidsun.atlassian.net/wiki/spaces/ELEC/pages/2030895112/Bootloader+Flash+over+CAN) for more info):

Effectively, there is one state machine which is used for both rx and tx. 

For tx, message are sent sequentially through the state machine as fast as they are processed. `can_datagram_start_tx()` starts the process, calling the tx callback with start=true, and raising the first state machine event.

For rx, an init message must be recieved  for processing to start. messages are received into a circular buffer, and then processed with sequential calls to can_datagram_rx.

To read/write datagram data, you must pass a buffer either with the relevant information, or with enough room to capture it.

### CODE REVIEW

**Review by AveryChiu64 — COMMENTED** [review]
Great work! This is clean. Though, it looks like the build is failing due to some formatting issues. Could you address this?

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [reply libraries/ms-common/src/can_datagram.c:103]: Yeah good point. I added a catch to exit with an error. 
We can't really add any other failsafe to wait for read to happen I thinkl, as the point of the buffer itself was to do this, so failing out should be enough

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [reply libraries/ms-common/src/can_datagram.c:69]: DST is the notation they use for "Destination nodes", which is the ids of all the nodes the message needs to reach. `dst_len` is essentially the number of uint8 node ids the datagram has. It is definitely not the most clear, I could change 'dst' to 'dest' or 'nodes' if that's clearer, or I could just add some comments?

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:385]: The API for this module is a little weird - to send/receive multiple messages you'd have to reinitialize it! Here is my idea for how to make the API more intuitive:
- make `can_datagram_init` a thin init function that only initializes the fsm and status and stuff like that
- give `state_done` a way to return to `state_idle` once we're done rxing/txing a datagram
- have `can_datagram_start_listening` and `can_datagram_start_tx` functions which can only be called in `state_idle`, and which start an RX or TX respectively
  - `can_datagram_start_listening` initializes everything for RX mode, and takes the data buffer where we put the datagram data
  - `can_datagram_start_tx` initializes everything for TX mode, takes all the datagram fields, and starts the TX
- also give users a way to know if we're currently in the middle of a tx/rx so they know if it's ok to start listening/txing
- set the event data in all the events (or at least the completion event) to the state we're in (tx/rx) so that downstream clients know if we've just txed or rxed
- document all this in the header!

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:280]: I think we should take in a node id and then return early from rxing any datagrams that don't have a destination node matching our node id since they aren't meant for us. That also might mean we don't have to store the destination nodes!

File: libraries/ms-common/test/test_can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/test/test_can_datagram.c:112]: Why the `{}` on all the case labels?

File: libraries/ms-common/inc/can_datagram.h
Comment (ryandancy) [inline libraries/ms-common/inc/can_datagram.h:60]: Some of these structs are only used in the implementation I believe, see if you can move any of them to the C file so they aren't polluting client code namespaces.

**Review by ryandancy — CHANGES_REQUESTED** [review]
Didn't look at the tests, but wow this is way more involved that I thought it would be. Looks awesome.

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [reply libraries/ms-common/src/can_datagram.c:157]: Could also just condense the metadata section to `memcpy(stream, &dt, sizeof(dt))` and then append everything in the same way? This would probably be cleaner but will be more opaque and use more byte space
```
static uint32_t prv_can_datagram_compute_crc(void) {
  uint8_t stream[sizeof(CanDatagram) + 2*sizeof(uint32_t)];
  uint8_t *write = stream;
  CanDatagram *dgram = &s_store.dgram;

  // CRC for data and dst_nodes
  uint32_t dst_crc = crc32_arr(dgram->destination_nodes, dgram->destination_nodes_len);
  uint32_t data_crc = crc32_arr(dgram->data, dgram->data_len);

  // Copy dgram with metadata to stream
  memcpy(stream, &dgram, sizeof(CanDatagram));
  write += sizeof(CanDatagram);
  memcpy(write, &data_crc, sizeof(uint32_t));
  write += sizeof(uint32_t);
  memcpy(write, &data_crc, sizeof(uint32_t));
  write += sizeof(uint32_t);
  return crc32_arr(stream, (size_t)(write - stream));
  }
  ```

Comment (ryandancy) [reply libraries/ms-common/src/can_datagram.c:157]: Hm, with that I'm concerned that with the datagram not being packed the compiler might add padding in places that would be hard to replicate in Python - I like the original way because it's more explicit.

**Comment by mitchellostler** [comment]
Still working on it, just wanted to have my work in remote lol

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [inline libraries/ms-common/src/can_datagram.c:39]: @ryandancy With the new structure, the fifo where data is and process is 2048 bytes long. It is one instance which is used for both tx/rx (I am renaming it to s_buffer), and all received/transmitted data is processed through here. 

If I make it any smaller than 2048, then it overflows when the `data` segment received is large. This also has the added bonus of allowing me to read data and destination_node buffers in one call to fifo_pop_arr()  *See below*

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [inline libraries/ms-common/src/can_datagram.c:423]: @ryandancy (part 2) Because buffer is large enough to hold both the largest possible destination_nodes buffer (255) and  the largest possible data buffer (2048), we can read directly from the fifo in bulk, and eliminate a bunch of code.

I admit that this does seem a bit sus for `data`, as if it is the largest possible length it will fit the buffer exactly, but we can assume that for any given message, all other messages will be already read.

Tested it with max conditions and it is working, just wanted to clarify the idea here

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [inline libraries/ms-common/src/can_datagram.c:485]: This is valid, as the max length pushed to the fifo for tx is 8 bytes, so we won't lose any data

**Comment by mitchellostler** [comment]
Alrighty, I've updated the API, fixed the PR comments, and added a few more tests. Everything seems to be working.
It's worth noting though, that our watchdog library seems to be having some issues giving some weird behaviour, I'm not to sure what's up with that.

I will improve the explanation in the header and rewrite the SDD once the changes are approved 🥇

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [reply libraries/ms-common/src/can_datagram.c:531]: - [ ] make can_datagram_init a thin init function that only initializes the fsm and status and stuff like that

- [ ] give state_done a way to return to state_idle once we're done rxing/txing a datagram

- [ ] have can_datagram_start_listening and can_datagram_start_tx functions which can only be called in state_idle, and which start an RX or TX respectively

- [ ] can_datagram_start_listening initializes everything for RX mode, and takes the data buffer where we put the datagram data
    - Now using tx and rx config structs  

- [ ] can_datagram_start_tx initializes everything for TX mode, takes all the datagram fields, and starts the TX

also give users a way to know if we're currently in the middle of a tx/rx so they know if it's ok to start listening/txing
set the event data in all the events (or at least the completion event) to the state we're in (tx/rx) so that downstream clients know if we've just txed or rxed
   - there are now two different statuses used for tx/rx completion

File: libraries/ms-common/src/watchdog.c
Comment (mitchellostler) [reply libraries/ms-common/src/watchdog.c:23]: There may be some issues in the Watchdog changes or library in general, I was getting some weird behaviour. Not too sure what it is though

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:588]: Might want to specify in the header that the `CanDatagramRxConfig` pointer has to remain valid until the datagram is fully sent, but the `CanDatagramTxConfig` pointer for `can_datagram_start_tx` doesn't (although its `data` field does).

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:232]: Even though you've already null checked elsewhere, maybe surround this with a null check just in case :)

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:578]: What's the strategy for notifying client code when we've successfully received a datagram? Currently we seem to just notify by setting the status - maybe we should take an optional "rx success" callback? (And on that note - perhaps an optional "tx success" callback in `can_datagram_start_tx` and an optional "error" callback in `can_datagram_init` (or maybe both tx and start listener functions) for hard errors?)

File: libraries/ms-common/src/can_datagram.c
Comment (ryandancy) [inline libraries/ms-common/src/can_datagram.c:622]: Hm, I suppose this does conform to the standard - my original thought was that the start message would also contain data, so we'd just set the start flag to true on the first tx_cb call. That way is more efficient especially for short messages, but this is simpler - your call on whether to change it or not.

**Review by ryandancy — CHANGES_REQUESTED** [review]
I haven't looked through the tests but it looks great! Pretty much all of the comments are minor stuff that's quick to fix, thanks for getting this through.

File: libraries/ms-common/src/can_datagram.c
Comment (mitchellostler) [reply libraries/ms-common/src/can_datagram.c:359]: I realized I am handling the final messages twice, just do it at the end here now

**Review by ryandancy — COMMENTED** [review]
Looked it all over and it looks good, just need the endianness-independence change + whatever CRC change you're making

**Review by ryandancy — APPROVED** [review]

**Pull request merged**
